language/move-stdlib/src/natives/hash.rs

```diff
@@ -26,11 +34,7 @@ pub fn native_sha2_256(
 
     let hash_arg = pop_arg!(arguments, Vec<u8>);
 
-    let cost = native_gas(
-        context.cost_table(),
-        NativeCostIndex::SHA2_256,
-        hash_arg.len(),
-    );
+    let cost = gas_params.base_cost + gas_params.unit_cost * usize::max(hash_arg.len(), 1) as u64;
```

Comment:
A general question: am I allowed to change a gas formula so that there doesn't exist any configuration that would allow it to maintain full backward compatibility? In other words, can I make breaking changes?

We have `unit_cost * max(len, 1)` in a few places and I kinda want to replace them with `base_cost + unit_cost * len`

Reply:
I think it is fine for this case? As if it is an issue for anyone they can write their own native.

But we might want to just stick two versions of the function in this file to help people?

Reply:
I think in the short term, I can just add another gas parameter `min_len` to allow the new formula to be compatible with the old one.

This doesn't answer my original question though. We still need to decide come up with a general guideline/policy on how we handle breaking changes to the gas formulae.

Reply:
@jolestar Any comments on this? You guys are ahead of us in terms of running a production network and I wonder if you have any insights to share.

Reply:
If the gas formula cost != cost_table cost, will break the compatibility. But I think the formula approach is better than the cost table, so we look forward to seeing this change.

We can try to adjust the gas params and let the two costs be the same, or just fork the `native_sha2_256` and write some code like if chain height > xxx for compatibility.

Reply:
Sounds good. Also had a discussion with @davidiw and we came to the conclusion that we should treat gas formula changes just like any other breaking changes, try to maintain backward compatibility for a transient period at least. This is to allow node operators to update their rust binaries in an asynchronous manner.